### PR TITLE
feat(bigquery): do not send preview field and rows if max_preview_rows is -1

### LINKED ISSUE #43: Restrict sending PreviewRows fields in table asset from meteor to compass
**Problem statement**
Currently when **MaxPreviewRows** value in meteor recipe is **0**, meteor fetches 0 rows from BQ and send null value of **PreviewRows** in table asset. As compass upsert api is being called, it updates the preview rows in compass asset to null. But for keeping preview rows static in compass we do not have any config which doesn't update preview rows into compass and let it be static.

**Proposed solution**
Compass doesn't update the field which is not being sent into asset. Hence to solve this, we come up with the solution to not to send **PreviewRows** field in table asset when **MaxPreviewRows** is set to **-1**. Hence it will not update the PreviewRows in compass and make it static.

```
source:
  type: bigquery
  scope: sample-scope
  config:
    project_id: sample-project
    max_preview_rows: -1
```

**28-Nov-2023 [Update]**

Even after following the above approach in bigquery extractor, it is not completely solved.
Following code at sink side adds back the unpopulated fields with default values while marshalling back to proto.

```
	data, err := protojson.MarshalOptions{
		UseProtoNames:   true,
		EmitUnpopulated: true,
	}.Marshal(anyData)
```

EmitUnpopulates is set to **true** which adds the unpopulated fields with empty value i.e. null and [] for preview_rows and preview_fields
Hence finally empties value at compass side.

Following are the proposed solutions for this issue:  (Preferred approach is 2)

**Approach 1**

Hard code EmitUnpopulated to **false** in compass sink like following 
```
	data, err := protojson.MarshalOptions{
		UseProtoNames:   true,
		EmitUnpopulated: false,
	}.Marshal(anyData)
```

Cons: It will change this configuration for all fields in asset for every every compass sink where this logic may not be needed.

**Approach 2**

Create a backward compatible sink config i.e. **remove_unset_fields**  to either **true or false** which defaults to **false** and keep the negation of it in marshal options like following code

`	RemoveUnsetFields   bool mapstructure:"remove_unset_fields"
`

```
	data, err := protojson.MarshalOptions{
		UseProtoNames:   true,
		EmitUnpopulated: !s.config.RemoveUnsetFields,
	}.Marshal(anyData)
```

Pros: It will be configurable and backward compatible. i.e. when remove_unset_fields is not set , it will default to false and hence finally EmitUnpopulated to true, So no changes required for current deployments.

Cons: It is changing config for all fields available in asset data.

**Approach 3**

**Extra changes in BQ extractor:** Unmarshal table asset after removing preview_rows  and preview_fields as done above  and marshal it back to new proto not having PreviewRows and PreviewFields (Maybe TableV2). Hence final proto would be marshal at sink end according to the new proto sent i.e TableV2 having no unset fields.

Pros: No changes required at sink side
Cons: Extra cost and time to convert and back to new proto, hence extra processing.

### PR DESCRIPTION
**Pull request by utsav14nov**
closes #43

### CODE REVIEW

**Review by anjali9791 — APPROVED** [review]

File: plugins/sinks/compass/sink.go [hunk 1]
```diff
@@ -28,9 +28,10 @@ import (
 var summary string
 
 type Config struct {
-	Host    string            `mapstructure:"host" validate:"required"`
-	Headers map[string]string `mapstructure:"headers"`
-	Labels  map[string]string `mapstructure:"labels"`
+	Host              string            `mapstructure:"host" validate:"required"`
+	Headers           map[string]string `mapstructure:"headers"`
+	Labels            map[string]string `mapstructure:"labels"`
+	RemoveUnsetFields bool              `mapstructure:"remove_unset_fields" default:"false"`
```
Comment (bsushmith) [inline plugins/sinks/compass/sink.go]: As discussed, Can you rename this field to denote its in asset/request something like - `RemoveUnsetFieldsInAsset`

Comment (utsav14nov) [reply plugins/sinks/compass/sink.go]: Changed it to `RemoveUnsetFieldsInData` as that looks more generic

File: plugins/extractors/bigquery/bigquery.go [hunk 2]
```diff
@@ -515,7 +521,7 @@ func (e *Extractor) buildColumn(ctx context.Context, field *bigquery.FieldSchema
 
 func (e *Extractor) buildPreview(ctx context.Context, t *bigquery.Table, md *bigquery.TableMetadata) (fields []string, rows *structpb.ListValue, err error) {
 	maxPreviewRows := e.config.MaxPreviewRows
-	if maxPreviewRows == 0 {
+	if maxPreviewRows == 0 || maxPreviewRows == -1 {
```
Comment (bsushmith) [inline plugins/extractors/bigquery/bigquery.go]: Can this be simply 
```suggestion
	if maxPreviewRows <= 0 {
```

File: plugins/extractors/bigquery/bigquery.go [hunk 3]
```diff
@@ -444,15 +444,21 @@ func (e *Extractor) buildAsset(ctx context.Context, t *bigquery.Table, md *bigqu
...
+	tableData := &v1beta2.Table{
+		Columns:    e.buildColumns(ctx, md.Schema, md),
+		Profile:    tableProfile,
+		Attributes: utils.TryParseMapToProto(attributesData),
+		CreateTime: timestamppb.New(md.CreationTime),
+		UpdateTime: timestamppb.New(md.LastModifiedTime),
+	}
+
+	maxPreviewRows := e.config.MaxPreviewRows
+	if maxPreviewRows != -1 {
```
Comment (bsushmith) [inline plugins/extractors/bigquery/bigquery.go:457]: Can you add documenation for this field and value in the struct and also in readme. 
struct - https://github.com/goto/meteor/blob/main/plugins/extractors/bigquery/bigquery.go#L52
readme - https://github.com/goto/meteor/blame/main/plugins/extractors/bigquery/README.md#L52

File: plugins/extractors/bigquery/README.md [hunk 4]
```diff
@@ -49,7 +50,7 @@ source:
 | `table_pattern` | `string` | `gofood.fact_` | Regex pattern to filter which bigquery table to scan (whitelist) | *optional* |
 | `max_page_size` | `int` | `100` | max page size hint used for fetching datasets/tables/rows from bigquery | *optional* |
 | `include_column_profile` | `bool` | `true` | true if you want to profile the column value such min, max, med, avg, top, and freq | *optional* |
-| `max_preview_rows` | `int` | `30` | max number of preview rows to fetch, `0` will skip preview fetching. Default to `30`. | *optional* |
+| `max_preview_rows` | `int` | `30` | max number of preview rows to fetch, `0` will skip preview fetching, `-1` will restruct adding preview_rows key in asset data . Default to `30`. | *optional* |
```
Comment (bsushmith) [inline plugins/extractors/bigquery/README.md]:
```suggestion
| `max_preview_rows` | `int` | `30` | max number of preview rows to fetch, `0` will skip preview fetching, `-1` will restrict adding preview_rows key in asset data . Default to `30`. | *optional* |
```

**Review by bsushmith — APPROVED** [review]

**Pull request merged**